### PR TITLE
feat: client-discord stop implementation / agent improvements

### PR DESCRIPTION
**Pull request by odilitime**
# Risks

Low

# Background

## What does this PR do?

implements discord websocket disconnect
make character settings optional
implement runtime.clients as a dict
disable client in characters client

## What kind of change is this?

Improvements (misc. changes to existing features)

## Why are we doing this? Any context or related work?

Preparation for hot reloading

# Documentation changes needed?

My changes should not require a change to the project documentation.

### CODE REVIEW

File: agent/src/index.ts [hunk 1]
```diff
@@ -318,42 +318,53 @@ function initializeDatabase(dataDir: string) {
...
+// also adds plugins from character file into the runtime
 export async function initializeClients(
     character: Character,
     runtime: IAgentRuntime
 ) {
-    const clients = [];
-    const clientTypes =
+    // each client can only register once
+    // and if we want two we can explicitly support it
+    const clients:Record<string, any> = {};
```
Comment (monilpat) [inline agent/src/index.ts:328]: Thanks so much for doing this! Appreciate how hard you are grinding! Looks like a space is missing after the colon for all the type declarations here

File: agent/src/index.ts [hunk 2]
```diff
@@ -318,42 +318,53 @@ function initializeDatabase(dataDir: string) {
...
-        clients.push(twitterClients);
+        const twitterClient = await TwitterClientInterface.start(runtime);
+        if (twitterClient) clients.twitter = twitterClient;
     }
 
     if (clientTypes.includes("farcaster")) {
-        const farcasterClients = new FarcasterAgentClient(runtime);
-        farcasterClients.start();
-        clients.push(farcasterClients);
+        // why is this one different :(
```
Comment (monilpat) [inline agent/src/index.ts:355]: We could create an issue for this to standardize

File: packages/core/src/types.ts [hunk 3]
```diff
@@ -603,7 +603,9 @@ export type Plugin = {
  */
 export enum Clients {
     DISCORD = "discord",
-    DIRECT = "direct",
+// you can't specify this in characters
+// all characters are registered with this
+//    DIRECT = "direct",
```
Comment (monilpat) [inline packages/core/src/types.ts:608]: not sure if this was intentional

**Review by monilpat — APPROVED** [review]
Overall LGTM - left a few nits nothing blocking - great work!

**Comment by monilpat** [comment]
Also addressing linter and then good to go thanks so much!

File: packages/core/src/types.ts (hunk 3 above)
Comment (odilitime) [reply packages/core/src/types.ts:608]: it is. Direct is meant to always run now

**Pull request merged**